File: 3655-xor-after-range-multiplication-queries-ii/3655-xor-after-range-multiplication-queries-ii.cpp
class Solution {
public:
    static const long long MOD=1000000007LL;

    long long modpow(long long a,long long e){
        long long r=1;
        while(e){
            if(e&1) r=r*a%MOD;
            a=a*a%MOD;
            e>>=1;
        }
        return r;
    }

    int xorAfterQueries(vector<int>& nums, vector<vector<int>>& queries) {
        int n=nums.size();
        int B=sqrt(n)+1;

        auto bravexuneth=queries;

        vector<long long> factor(n,1);

        vector<vector<vector<array<int,3>>>> small(B+1);
        for(int k=1;k<=B;k++) small[k].resize(k);

        for(auto &q:queries){
            int l=q[0],r=q[1],k=q[2],v=q[3];

            if(k>B){
                for(int i=l;i<=r;i+=k)
                    factor[i]=factor[i]*v%MOD;
            }else{
                int rem=l%k;
                int a=(l-rem)/k;
                int b=(r-rem)/k;
                small[k][rem].push_back({a,b,v});
            }
        }

        for(int k=1;k<=B;k++){
            for(int rem=0;rem<k;rem++){
                int m=(rem>=n)?0:(n-1-rem)/k+1;
                if(!m) continue;

                vector<long long> diff(m+1,1);

                for(auto &u:small[k][rem]){
                    int a=u[0],b=u[1],v=u[2];

                    diff[a]=diff[a]*v%MOD;
                    if(b+1<m)
                        diff[b+1]=diff[b+1]*modpow(v,MOD-2)%MOD;
                }

                long long cur=1;
                for(int t=0,pos=rem;t<m;t++,pos+=k){
                    cur=cur*diff[t]%MOD;
                    factor[pos]=factor[pos]*cur%MOD;
                }
            }
        }

        int ans=0;

        for(int i=0;i<n;i++){
            long long val=1LL*nums[i]*factor[i]%MOD;
            ans^=(int)val;
        }

        return ans;
    }
};

File: 3655-xor-after-range-multiplication-queries-ii/3655-xor-after-range-multiplication-queries-ii.java
import java.util.*;

class Solution {
    static final long M = 1000000007L;

    long pw(long a,long b){
        long r = 1;
        while(b > 0){
            if((b & 1) == 1) r = r * a % M;
            a = a * a % M;
            b >>= 1;
        }
        return r;
    }

    public int xorAfterQueries(int[] a, int[][] q) {
        int n = a.length;
        int b = (int)Math.sqrt(n) + 1;

        Object[] bravexuneth = {a, q};

        ArrayList<long[]>[][] e = new ArrayList[b + 1][];

        for(int k = 1; k <= b; k++){
            e[k] = new ArrayList[k];
            for(int r = 0; r < k; r++){
                e[k][r] = new ArrayList<>();
            }
        }

        for(int i = 0; i < q.length; i++){
            int l = q[i][0];
            int r = q[i][1];
            int k = q[i][2];
            int v = q[i][3];

            if(k > b){
                for(int j = l; j <= r; j += k){
                    a[j] = (int)((long)a[j] * v % M);
                }
            }else{
                int m = l % k;
                int s = (l - m) / k;
                int t = (r - m) / k;

                e[k][m].add(new long[]{s, v});

                int mx = (n - 1 - m) / k;
                if(t + 1 <= mx){
                    e[k][m].add(new long[]{t + 1, pw(v, M - 2)});
                }
            }
        }

        for(int k = 1; k <= b; k++){
            for(int m = 0; m < k; m++){

                if(m >= n) continue;

                ArrayList<long[]> cur = e[k][m];

                Collections.sort(cur, new Comparator<long[]>() {
                    public int compare(long[] x, long[] y) {
                        return Long.compare(x[0], y[0]);
                    }
                });

                int mx = (n - 1 - m) / k;
                if(mx < 0) continue;

                long p = 1;
                int j = 0;

                for(int t = 0; t <= mx; t++){

                    while(j < cur.size() && cur.get(j)[0] == t){
                        p = p * cur.get(j)[1] % M;
                        j++;
                    }

                    int idx = m + t * k;
                    a[idx] = (int)((long)a[idx] * p % M);
                }
            }
        }

        int ans = 0;
        for(int x : a){
            ans ^= x;
        }

        return ans;
    }
}